core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -192,4 +209,65 @@ public synchronized T next() {
       return queue.poll();
     }
   }
+
+  private static class Task<T> implements Callable<Optional<Task<T>>>, Closeable {
+    private final Iterable<T> input;
+    private final ConcurrentLinkedQueue<T> queue;
+    private final AtomicBoolean closed;
+    private final int approximateMaxQueueSize;
+
+    private Iterator<T> iterator;
+
+    Task(
+        Iterable<T> input,
+        ConcurrentLinkedQueue<T> queue,
+        AtomicBoolean closed,
+        int approximateMaxQueueSize) {
+      this.input = Preconditions.checkNotNull(input, "input cannot be null");
+      this.queue = Preconditions.checkNotNull(queue, "queue cannot be null");
+      this.closed = Preconditions.checkNotNull(closed, "closed cannot be null");
+      this.approximateMaxQueueSize = approximateMaxQueueSize;
+    }
+
+    @Override
+    public Optional<Task<T>> call() throws Exception {
+      try {
+        if (iterator == null) {
+          iterator = input.iterator();
+        }
+        while (iterator.hasNext()) {
+          if (queue.size() >= approximateMaxQueueSize) {
+            // yield
+            return Optional.of(this);
```

Comment:
If processing of the items from the queue is slower, this may mean that in `checkTasks()` we're processing the same item over and over again.


From `checkTasks()`

```
          if (taskFutures[i] != null) {
            Optional<Task<T>> continuation;
            // check for task failure and re-throw any exception
            try {
              continuation = taskFutures[i].get();
            } catch (...)
                ...
            }
            continuation.ifPresent(yieldedTasks::addLast);
          }

          taskFutures[i] = submitNextTask();
```

If we know that we're dealing with a "hold your horses" task, why do we re-submit it again?

Reply:
i am not sure i follow. do you mean something is not correct?

Reply:
> i am not sure i follow. 

Why do we potentially keep on re-submitting the task  ?

I'm saying why not setting `taskFutures[i] = continuation.get()` ?

Reply:
> I'm saying why not setting `taskFutures[i] = continuation.get()` ?

But `continuation.get()` is of type `Optional<Task>`, while `taskFutures[i]` is of type `Future<Optional<Task>>`, so how would this compile?

> this may mean that in checkTasks() we're processing the same item over and over again.

We may be processing the same *task* but not the same item.  Each time a task is submitted, it grabs new items from its iterable and puts them into the queue.  Prior to this patch, the task goes over the entire iterable uninterrupted, aggressively growing the queue without any limits.  That's where the memory pressure that users are complaining about comes from.

With this patch, the task stops when the queue size reaches `approximateMaxQueueSize` and remembers where it stopped in internal state.  When it gets resubmitted later, it resumes filling the queue until the limit is reached again, then stops again, and so on.  Because tasks stop voluntarily and yield the control back to the executor, they do not block other ongoing threads from the same pool.

Reply:
That's a good explanation, thank you @dekimir !

> re-submitting the task

Worth noting that it's impl detail of the Task class that this is the same task.
It could be something else. It's just "a continuation". The task says "i did what i did, i am giving up now, and here, in the return value is the continuation that will do the rest of the work".